### DISCUSSION TITLE
Export as Premiere Pro .prtranscript file

### DISCUSSION THREAD

**Discussion by caleidoscopio** [openai/whisper#1073] [Ideas]
It would be super useful to export Whisper transcripts as .prtranscript files to import in Premiere Pro. 

A .prtranscript file differs from an .srt by having a timecode for each word, rather that each line, essentially.

This would make it possible to do text based editing in languages where Premiere Pro's transcription is not that good.

Here's an example of how text based editing works:
https://youtu.be/O_CFEsOPCkY?t=64

**Comment by glangford** [comment]
Is the .prtranscript file format documented anywhere, or is it proprietary?

**Reply by caleidoscopio** [reply]
Probably proprietary.

**Comment by salomaoparkour** [comment]
This would be perfect!!

**Comment by ytcyde** [comment]
Seems hard to implement considering this is how a prtranscript file looks like in a text editor.
![image](https://github.com/user-attachments/assets/66b40f84-4aa6-4819-84d2-de77469bcee4)

**Comment by DeL7266** [comment]
Kick me when you guys solve this problem. I have the same issue

**Comment by IzoFox** [comment]
Yeah, it would be good to make something like this.

**Comment by hisprofile** [comment]
please...

**Comment by miambe** [comment]
@caleidoscopio @hisprofile @IzoFox @DeL7266 @ytcyde @salomaoparkour @glangford it's easy
I already have a working implementation that can parse and convert Adobe Premiere Pro `.prtranscript` files without difficulty. The binary format, its FlatBuffers structure, the speaker and segment tables, the word-level timing data, and Adobe tick arithmetic are therefore already understood and implemented on my side.

The missing feature is the reverse operation: generating a valid `.prtranscript` file directly from a Whisper transcription.

This is technically feasible, but exporting a Whisper transcript as a `.prtranscript` file requires more than simply converting an SRT file.

A Premiere Pro `.prtranscript` file contains word-level timestamps, speaker information, language metadata, confidence scores, and token types. Whisper can provide most of the transcription data required, especially when word-level timestamps are enabled. The remaining step is to serialize that information into the binary structure expected by Adobe Premiere Pro.

I have reverse-engineered the relevant parts of that format, and the existing parser provides a reliable basis for implementing the corresponding writer.


Exporting a Whisper transcript as a <code inline="">.prtranscript</code> file requires more than simply converting an SRT file.</p><p>A Premiere Pro <code inline="">.prtranscript</code> file contains word-level timestamps, speaker information, language metadata, confidence scores, and token types. Whisper can provide most of the transcription data required, especially when word-level timestamps are enabled. The remaining step is to serialize that information into Adobe Premiere Pro’s proprietary binary format.</p><p>I have reverse-engineered the relevant parts of that format.</p><h2>1. Proposed Workflow</h2><p>A possible export pipeline would be:</p><pre><code class="language-text">Audio or video file
        ↓
Whisper transcription with word-level timestamps
        ↓
Optional speaker diarization
        ↓
Transcript normalization
        ↓
Adobe tick conversion
        ↓
FlatBuffers serialization
        ↓
.prtranscript file
</code></pre><p>Unlike an <code inline="">.srt</code> file, which usually assigns one time range to an entire subtitle line, a <code inline="">.prtranscript</code> file stores timing information for every individual word.</p><p>This would indeed make Premiere Pro’s text-based editing available for languages where Adobe’s built-in transcription is less accurate than Whisper.</p><h2>2. Binary Specification of the Proprietary <code inline="">.prtranscript</code> Format</h2><p>Adobe Premiere Pro <code inline="">.prtranscript</code> files are binary buffers serialized using the <strong>FlatBuffers</strong> protocol.</p><p>Unlike JSON or XML, FlatBuffers allows applications to access internal fields through relative offsets without fully deserializing the entire document.</p><p>The following specification is reverse-engineered and is not based on an official Adobe schema. Adobe may therefore change the format in future Premiere Pro versions.</p><h3>2.1 File Memory Layout</h3><pre>
<code class="language-mermaid">classDiagram
    class FlatBufferFile {
        +u32 root_table_offset
        +RootTable root_table
    }

    class RootTable {
        +i32 vtable_offset_diff
        +u32 root_data_offset
    }

    class RootData {
        +u32 segments_vector_offset
        +u32 language_string_offset
        +u32 speakers_vector_offset
    }

    class SpeakerTable {
        +u32 name_offset
        +u32 uuid_table_offset
    }

    class SegmentTable {
        +i64 start_ticks
        +i64 duration_ticks
        +u32 speaker_uuid_offset
        +u32 words_vector_offset
        +u32 language_offset
    }

    class WordTable {
        +i64 start_ticks
        +i64 duration_ticks
        +u32 text_offset
        +f32 confidence
        +u8 token_type
    }

    FlatBufferFile --&gt; RootTable
    RootTable --&gt; RootData
    RootData --&gt; SpeakerTable : Vector
    RootData --&gt; SegmentTable : Vector
    SegmentTable --&gt; WordTable : Vector
</code></pre><h3>2.2 Offsets and Table Structure</h3><ol><li><p><strong>Header — bytes 0 through 3</strong></p><p>The first four bytes contain a little-endian <code inline="">u32</code> value pointing to the beginning of the root table.</p></li><li><p><strong>VTable</strong></p><p>Each FlatBuffers table begins with a signed offset pointing backwards to its associated virtual table, or <code inline="">vtable</code>.</p><p>The <code inline="">vtable</code> contains the relative offsets of the fields that are present in the serialized object. Missing optional fields normally have an offset of zero.</p></li><li><p><strong><code inline="">RootData</code> table</strong></p><p>The root data table contains the document-level transcript information:</p><ul><li><p><strong>Field 0:</strong> Offset to the vector of <code inline="">SegmentTable</code> objects.</p></li><li><p><strong>Field 1:</strong> Offset to the document language string, for example <code inline="">"fr-fr"</code>.</p></li><li><p><strong>Field 2:</strong> Offset to the vector of <code inline="">SpeakerTable</code> objects.</p></li></ul></li><li><p><strong><code inline="">SpeakerTable</code></strong></p><p>Each speaker entry contains:</p><ul><li><p><strong>Field 0:</strong> Offset to the speaker name string.</p></li><li><p><strong>Field 1:</strong> Offset to a 128-bit UUID table.</p></li></ul><p>The UUID is stored as two 64-bit integer values in fields 0 and 1 of the nested UUID table.</p></li><li><p><strong><code inline="">SegmentTable</code></strong></p><p>Each transcript segment contains:</p><ul><li><p><strong>Field 0:</strong> Segment start time as an <code inline="">i64</code> Adobe tick value.</p></li><li><p><strong>Field 1:</strong> Segment duration as an <code inline="">i64</code> Adobe tick value.</p></li><li><p><strong>Field 2:</strong> Offset to the UUID table of the associated speaker.</p></li><li><p><strong>Field 3:</strong> Offset to the vector of <code inline="">WordTable</code> objects.</p></li><li><p><strong>Field 4:</strong> Optional segment-specific language string.</p></li></ul></li><li><p><strong><code inline="">WordTable</code></strong></p><p>Each word or token contains:</p><ul><li><p><strong>Field 0:</strong> Word start time as an <code inline="">i64</code> Adobe tick value.</p></li><li><p><strong>Field 1:</strong> Word duration as an <code inline="">i64</code> Adobe tick value.</p></li><li><p><strong>Field 2:</strong> Offset to the word text string.</p></li><li><p><strong>Field 3:</strong> Optional confidence score stored as an <code inline="">f32</code>.</p></li><li><p><strong>Field 4:</strong> Optional token type stored as a <code inline="">u8</code>, although some implementations may expose it as an integer.</p></li></ul></li></ol><p>A normalized intermediate representation could look like this:</p><pre><code class="language-json">{
  "language": "fr-fr",
  "speakers": [
    {
      "id": "8e8ac2a1-2ea2-4708-9bc8-d31e884752af",
      "name": "Speaker 1"
    }
  ],
  "segments": [
    {
      "start": 9.36,
      "end": 22.06,
      "speaker_id": "8e8ac2a1-2ea2-4708-9bc8-d31e884752af",
      "language": "fr-fr",
      "words": [
        {
          "text": "You",
          "start": 9.51,
          "end": 9.66,
          "confidence": 0.98,
          "token_type": 1
        }
      ]
    }
  ]
}
</code></pre><p>That intermediate JSON can then be encoded into the FlatBuffers structure expected by Premiere Pro.</p><h2>3. Timestamp Arithmetic: Adobe Ticks to Milliseconds</h2><p>Premiere Pro expresses timestamps and durations using high-precision <strong>Adobe ticks</strong>.</p><pre><code class="language-text">1 second = 254,016,000,000 ticks
</code></pre><p>This value can be defined as:</p><pre><code class="language-rust">pub const ADOBE_TICKS_PER_SECOND: u64 = 254_016_000_000;
</code></pre><h3>3.1 Exact Conversion Formula</h3><p>To avoid precision loss caused by floating-point arithmetic, timestamps should be converted using exact integer arithmetic.</p>

The conversion to milliseconds is:

```text
T_ms = floor((T_ticks * 1000) / 254,016,000,000)
```

The inverse conversion is:

```text
T_ticks = floor((T_ms * 254,016,000,000) / 1000)
```
<p>For Whisper timestamps expressed in seconds, it is preferable to avoid directly multiplying an imprecise <code inline="">f64</code> value when exact frame alignment matters.</p><p>A practical implementation can first convert the timestamp into integer microseconds or nanoseconds and then convert that value into Adobe ticks.</p><h3>3.2 Rust Implementation</h3><pre><code class="language-rust">pub const ADOBE_TICKS_PER_SECOND: u64 = 254_016_000_000;

pub fn ticks_to_milliseconds(
    ticks: i64,
) -&gt; Result&lt;u64, PrTranscriptError&gt; {
    // Reject negative timestamps.
    let ticks_u128 = u128::try_from(ticks)
        .map_err(|_| PrTranscriptError::NegativeTimestamp { ticks })?;

    // Perform the multiplication and division using u128
    // to avoid intermediate overflows.
    let milliseconds = ticks_u128
        .checked_mul(1000)
        .ok_or(PrTranscriptError::TimestampOverflow)?
        / ADOBE_TICKS_PER_SECOND as u128;

    u64::try_from(milliseconds)
        .map_err(|_| PrTranscriptError::TimestampOverflow)
}
</code></pre><p>The reverse conversion can be implemented as follows:</p><pre><code class="language-rust">pub fn milliseconds_to_ticks(
    milliseconds: u64,
) -&gt; Result&lt;i64, PrTranscriptError&gt; {
    let ticks = (milliseconds as u128)
        .checked_mul(ADOBE_TICKS_PER_SECOND as u128)
        .ok_or(PrTranscriptError::TimestampOverflow)?
        / 1000;

    i64::try_from(ticks)
        .map_err(|_| PrTranscriptError::TimestampOverflow)
}
</code></pre><p>For higher timing precision, a microsecond-based conversion is preferable:</p><pre><code class="language-rust">pub fn microseconds_to_ticks(
    microseconds: u64,
) -&gt; Result&lt;i64, PrTranscriptError&gt; {
    let ticks = (microseconds as u128)
        .checked_mul(ADOBE_TICKS_PER_SECOND as u128)
        .ok_or(PrTranscriptError::TimestampOverflow)?
        / 1_000_000;

    i64::try_from(ticks)
        .map_err(|_| PrTranscriptError::TimestampOverflow)
}
</code></pre><h2>4. Whisper Data Mapping</h2><p>The principal Whisper fields can be mapped as follows:</p>
Whisper data | .prtranscript field
-- | --
Word start time | WordTable.start_ticks
Word end minus start | WordTable.duration_ticks
Word text | WordTable.text
Word confidence | WordTable.confidence
Segment start time | SegmentTable.start_ticks
Segment end minus start | SegmentTable.duration_ticks
Detected language | Root or segment language
Diarized speaker | Speaker UUID
Punctuation or token category | WordTable.token_type

<p>If the Whisper implementation does not provide speaker diarization, all segments can initially be assigned to a single generated speaker.</p><p>If confidence values are unavailable, the field can either be omitted or assigned a neutral default value, depending on what Premiere Pro accepts.</p><h2>5. Main Implementation Challenges</h2><p>The difficult part is not obtaining word-level timestamps from Whisper. The main challenges are:</p><ul><li><p>reproducing Adobe’s exact FlatBuffers schema;</p></li><li><p>generating UUIDs using the expected byte order;</p></li><li><p>correctly classifying punctuation and token types;</p></li><li><p>preserving gaps and silent tokens when Premiere expects them;</p></li><li><p>validating that the resulting file is accepted by different Premiere Pro versions;</p></li><li><p>handling possible schema changes introduced by Adobe.</p></li></ul><p>A robust implementation should therefore include round-trip tests against genuine <code inline="">.prtranscript</code> files exported by Premiere Pro.</p><h2>Conclusion</h2><p>A Whisper-to-<code inline="">.prtranscript</code> exporter is entirely achievable.</p><p>The core transcription information already exists in Whisper’s word-level output. The exporter mainly needs to normalize the transcript, convert timestamps into Adobe ticks, associate words with speakers and segments, and serialize the result using the reverse-engineered FlatBuffers structure expected by Premiere Pro.</p>

**Comment by jokeum** [comment]
Has anyone tried this official JSON spec format that supports word-level timecode for captioning and Text-Based Editing?
https://community.adobe.com/announcements-732/now-in-beta-import-your-own-transcript-311910
